### ISSUE TITLE
Make message transport extensible/pluggable

### ISSUE THREAD

**Issue by jannfis** [argoproj-labs/argocd-agent#260]
**Is your feature request related to a problem? Please describe.**

Quite a few things within argocd-agent are designed in a way that they can be easily replaced by a different implementation. Notable examples are authentication methods and resource backends for the principle.

However, one thing that's pretty much hard coded is the message exchange transport, which is based off a gRPC protocol. In some cases, gRPC may not be the favorable message transport. Or in some cases, it might not even be possible to use gRPC between the agents and the principle for reasons of security, governance or others and users might be forced to use messaging systems such as Kafka or other third party message routers.

**Describe the solution you'd like**

The internal message transport system should be redesigned so messages can be transported using arbitrary mechanisms, while gRPC will stay the default transport in the spirit of "batteries included".

**Describe alternatives you've considered**

**Additional context**

**Comment by mikeshng** [comment]
@bianbbc87 Thank you for taking an interest in tackling this task! Since its quite a large one, I would recommend starting by creating a publicly accessible Google Doc with more detailed design (maybe with some diagrams). That way we can review and comment on it, ensuring we are all on the same page. WDYT? @jannfis do you agree as well?

CC @jnpacker @xiangjingli

**Comment by jannfis** [comment]
I agree, @mikeshng. We should think this through thoroughly and come up with some concise requirements before heading into an implementation.

Instead of a public Google doc (which could imho be somewhat discriminating users who want to opt-out of Google) , we could also use a PR of a  markdown document similar to the ones in https://github.com/argoproj-labs/argocd-agent/tree/main/docs/technical to set out requirements and have the discussions on comments on the pull request.

**Comment by bianbbc87** [comment]
> I agree, [@mikeshng](https://github.com/mikeshng). We should think this through thoroughly and come up with some concise requirements before heading into an implementation.
> 
> Instead of a public Google doc (which could imho be somewhat discriminating users who want to opt-out of Google) , we could also use a PR of a markdown document similar to the ones in https://github.com/argoproj-labs/argocd-agent/tree/main/docs/technical to set out requirements and have the discussions on comments on the pull request.

Thank you for the feedback. I’ll prepare an initial design document as a PR under docs/technical and join the discussion there. Looking forward to your comments and suggestions.

CC @mikeshng

**Comment by arielrahamim** [comment]
We deploy K3s edge clusters at customer premises, some behind Forcepoint SSL-inspecting proxies. From operating these clusters we've learned that Forcepoint blocks gRPC/HTTP2 (ALPN stripped), WebSocket (`Upgrade` header rejected), and non-standard TLS patterns. The only traffic that passes reliably is standard HTTPS/1.1 GET/POST. Because of these constraints, we design for the worst case and assume all customers won't have direct internet connectivity.

We ran an argocd-agent POC (v0.4.0, managed mode) over Istio mTLS through an east-west gateway — that works for customers with direct internet access. But for customers behind SSL-inspecting proxies there's no Istio, and all three current transports (gRPC/HTTP2, WebSocket, Istio mTLS) would be blocked.

An HTTP/1.1 long-poll transport would unlock agent support for these environments. Happy to share more details about the proxy behaviors we've observed if useful for the design.

**Comment by jannfis** [comment]
Thanks for these insights, @arielrahamim.

A scenario like you describe is definitely something the project would like to support in the future, so that we can be agnostic of any intermediaries in between the path from the agents to the principal.

Appreciate your offer to share more details, too.
